### PR TITLE
Enactstozarr

### PR DESCRIPTION
**Pull request by drewmresnick**
improving the file for enacts zarr conversion

### CODE REVIEW

File: enacts/enactstozarr.py
Comment (drewmresnick) [inline enacts/enactstozarr.py:43]: @remicousin not sure if I was to keep this regridding code in the function since you had not originally included in the function you started working on. I put it in there for now. Mybe we can make it a parameter that can be defined so you can choose if you want to use the regridding or not?

File: enacts/config-nma.yaml
Comment (aaron-kaplan) [inline enacts/config-nma.yaml:35]: Maybe we should we just mirror the structure of the original ENACTS dataset, so that element 0 and element 1 of each config are the same and can be combined? I imagine someone had a reason to add those extra directory levels.

File: enacts/enactstozarr.py
Comment (aaron-kaplan) [inline enacts/enactstozarr.py:14]: Rename this ZARR_RESOLUTION to avoid confusion?

File: enacts/config-nma.yaml
Comment (aaron-kaplan) [inline enacts/config-nma.yaml:30]: I don't understand why there are two zarr paths. Is that what you meant by your comment "this I could delete..." above? If rr_mrg_zarr_path is unused then definitely get rid of it.

File: enacts/config-nma.yaml
Comment (aaron-kaplan) [inline enacts/config-nma.yaml:31]: Call this `vars` or `variables` instead of `data_src`?

File: enacts/config-nma.yaml
Comment (drewmresnick) [reply enacts/config-nma.yaml:35]: I added the r0375 directory level because when the previous script was run it was run at a different resolution, so I thought it would make sense to have the option to create other directories if someone chooses to rerun the script at a different resolution like I did. As for the rr_mrg_ALL directory level, I cannot say. I just kept that because the previous path included it.

**Comment by drewmresnick** [comment]
@aaron-kaplan @remicousin once it is confirmed that the structure I proposed in the config-nma.yaml file for the vars dict is okay, then I will go back in and update all of the config files with that structure and will also go and update all the maproom files where the old file path configuration def is found to reflect the change. I just did not want to touch them until I know that the structure makes the most sense.

File: enacts/config-nma.yaml
Comment (aaron-kaplan) [reply enacts/config-nma.yaml:35]: Oh, I had it backwards. I thought the one with r0375 in it was the original. I agree with having subdirs for different zarr resolutions.

Not a fan of rr_mrg_ALL. I don't even know what the rr means.

**Comment by aaron-kaplan** [comment]
At some point we should come up with a coherent specification of a full ENACTS 
 site, including both ALL and MON, and covering multiple versions. ENACTS methodology seems to be pretty well standardized, but for whatever reason, the directory naming patterns are not, so it would be good to have a standard dataset structure that our maprooms all use internally, and then the job of the config file would be to map from the idiosyncratic directory structure of an individual ENACTS instance and the standard internal data structure.

This doesn't feel particularly urgent to me, though. I'm fine with the structure you've made here for now. I'm not sure what @remicousin had in mind when he asked you to work on this so I'll wait for him to comment.

**Comment by remicousin** [comment]
The goal of this PR is to make enactstozarr (re)usable again. My work on the wat_bal_monit maproom triggered rationalization of multiple copies of enactstozarr (among other things) which I did poorly as documented in this issue #253 

Historically, we try to be careful with our Data Cataloguing structure since: i) it's public facing; ii) it's what is ultimately invoked by Ingrid programs and iii) we have control of that. But not that careful with structure of how data files are organized. With Python Maprooms invoking data files directly and without an intermediary catalog, that frontier has vanished. And in order to keep track of things, we have had zarr directory structure correspond to data files directory structure (thus the less cautiously curated organization). Now, the resolution version 0.25 vs 0.0375 is not from ENACTS products but from us producing different zarr products; and thus by definition that resolution substructure is not part of the structure organizing ENACTS data files.

Long story short to say: it's a bit messy as enactstozarr involves a bunch of different things: nc files we read, zarr files we write, zarr files used by this or that app. And so as a consequence I unresolved the conversation about this and added a comment as the current proposition is probably not ok.

File: enacts/enactstozarr.py
Comment (aaron-kaplan) [reply enacts/enactstozarr.py:75]: I have something in progress that will let us incrementally add data to a zarr store as the upstream dataset gets extended in time, rather than re-converting the entire dataset on every run of the script. Once that's available, we'll rewrite this script in that paradigm; in the meantime, don't worry about it.

File: enacts/enactstozarr.py
Comment (aaron-kaplan) [reply enacts/enactstozarr.py:43]: Sure, making `resolution` a parameter of the `convert` function makes sense.

File: enacts/config-nma.yaml
Comment (aaron-kaplan) [reply enacts/config-nma.yaml:30]: I really think the path of the native resolution ENACTS data should be global, not page-specific. If certain pages need parts of their calculations done ahead of time for speed reasons (e.g. regridding in this case), those can be page-specific.

But my first thought here to wonder whether we really need to do the regridding ahead of time. The Ingrid version does the regridding on the fly, so maybe we can do that here too, and avoid the complexity of having multiple copies of the data at different resolutions. @drewmresnick are you up to trying that?

Comment (remicousin) [reply enacts/config-nma.yaml:30]: Then we should first try if no regridding at all is acceptable for the Ethiopia Onset maproom first. The regridding in Ingrid was only for performance issue. That test might have to be done on a machine more performant than devi to be relevant.

Comment (aaron-kaplan) [reply enacts/config-nma.yaml:30]: We did try the onset calculation at the native resolution, and that was too slow. There have been some other performance improvements since then, so I guess it's worth another try.

File: enacts/enactstozarr.py
Comment (drewmresnick) [reply enacts/enactstozarr.py:43]: Just to confirm, it would be a parameter where if it is set to true then it will compare the resolution such as in line 42, to attempt the regridding?

Comment (aaron-kaplan) [reply enacts/enactstozarr.py:43]: I imagined it would be a number. It could be optional--if it's not provided, then don't regrid. If it is provided, then check the resolution of the original dataset, and if it's not the same as the desired resolution, then regrid.

File: enacts/config-nma.yaml
Comment (drewmresnick) [reply enacts/config-nma.yaml]: I restructured / renamed the temp data to address Rémi's comment.

File: enacts/config-nma.yaml
Comment (drewmresnick) [reply enacts/config-nma.yaml:30]: I am unclear exactly what you are asking me to do @aaron-kaplan. Do you want me to work to create a function that can be applied to the data on-the-fly inside of a maproom instead of running the regridding on the data during conversion such as it is inside this file?

Comment (aaron-kaplan) [reply enacts/config-nma.yaml:30]: That's what I was suggesting, yes. But then Rémi made a different suggestion, which was to try not regridding at all, just use the dataset's native resolution in the onset maproom. I don't know if that will work, but it's easy to try, so let's try it.

File: enacts/enactstozarr.py
Comment (drewmresnick) [reply enacts/enactstozarr.py:75]: Yes and I guess my thought was someone could just comment out variables they wouldn't want for now.

**Comment by drewmresnick** [comment]
> The goal of this PR is to make enactstozarr (re)usable again. My work on the wat_bal_monit maproom triggered rationalization of multiple copies of enactstozarr (among other things) which I did poorly as documented in this issue #253
> 
> Historically, we try to be careful with our Data Cataloguing structure since: i) it's public facing; ii) it's what is ultimately invoked by Ingrid programs and iii) we have control of that. But not that careful with structure of how data files are organized. With Python Maprooms invoking data files directly and without an intermediary catalog, that frontier has vanished. And in order to keep track of things, we have had zarr directory structure correspond to data files directory structure (thus the less cautiously curated organization). Now, the resolution version 0.25 vs 0.0375 is not from ENACTS products but from us producing different zarr products; and thus by definition that resolution substructure is not part of the structure organizing ENACTS data files.
> 
> Long story short to say: it's a bit messy as enactstozarr involves a bunch of different things: nc files we read, zarr files we write, zarr files used by this or that app. And so as a consequence I unresolved the conversation about this and added a comment as the current proposition is probably not ok.

I am not exactly sure where this conversation leaves me in terms of what I am working on in this pull request.. are we to discuss further the regridding, and as a result also the file structure of the data?

File: enacts/enactstozarr.py
Comment (drewmresnick) [reply enacts/enactstozarr.py:43]: So then do you have in mind a different test than line 42?

`if not np.isclose(data['X'][1] - data['X'][0], ZARR_RESOLUTION):`

`ZARR_RESOLUTION` would be the number it would check against. So then if we are to use `ZARR_RESOLUTION` as the parameter I would think we would need to 1) probably rename that var because it is vague that this is what it is being used for and 2) be more specific as to what you would set `ZARR_RESOLUTION` as in the config file if you do not want to test for regridding.

Comment (aaron-kaplan) [reply enacts/enactstozarr.py:43]: The name of the parameter can just be `resolution`, and the configuration variable would be `CONFIG['onset']['resolution']`. Set it to `null` if you want to use the dataset's native resolution.

**Comment by remicousin** [comment]
I would say:
1/ move the current zarr data to a 0p25 subrepo (better to use 'p' to remove some ambiguity so r0375 should be r0p0375)
2/ create the native resolution zarr data without any resolution suprepo
3/ test onset maproom on that zarr data (maybe not on devi)
4/ if it seems to work ok: remove the regridding from the enactstozarr code
5/ if it doesn't: have the regridding done inside the onset maproom directly -- test performance of that
6/ if works ok. done
7/ if doesn't work ok: put back regridding option in enactstozarr code

yeah?

**Comment by aaron-kaplan** [comment]
> better to use 'p' to remove some ambiguity

Why not "." instead of "p"?

Otherwise we are in agreement.

**Comment by remicousin** [comment]
I am not sure if "." is too much of a special character but if not, that is all fine with me to use it.

**Comment by aaron-kaplan** [comment]
> I am not sure if "." is too much of a special character

Only in DOS and Windows prior to Windows 95, AFAIK.

**Comment by drewmresnick** [comment]
> I would say: 1/ move the current zarr data to a 0p25 subrepo (better to use 'p' to remove some ambiguity so r0375 should be r0p0375) 2/ create the native resolution zarr data without any resolution suprepo 3/ test onset maproom on that zarr data (maybe not on devi) 4/ if it seems to work ok: remove the regridding from the enactstozarr code 5/ if it doesn't: have the regridding done inside the onset maproom directly -- test performance of that 6/ if works ok. done 7/ if doesn't work ok: put back regridding option in enactstozarr code
> 
> yeah?

I tested the Zambia onset maproom using the regridded data and the data at the native resolution without any regridding. The runtime is not significantly greater (for example to return a new onset graph the difference is maybe 50-100 ms.) So it seems to me that it should work fine at the native resolution.

**Comment by remicousin** [comment]
I am a bit confused because Zambia didn't have a 0.25 version, only Ethiopia. Did you create one for your test? Also I would think the testing would matter slightly more on the time to map rather than the time to draw graphs... but maybe not...

Regardless, it seems you are all set to create the Ethiopia ENACTS 0.0375 zarr files and you should go ahead with that. I would put no extra directory in the folder path to reflect resolution if it's the native one. Then if someone really needs to create a regridded version, then they will add a   r0.25/ to their structure. And so, we can assume now that the Ethiopia onset Maproom will work ok at 0.0375 but Aaron should still add r.025/ to his zarr files path structure for the record and in case we need to revert back to it. And if we do then we'll revisit the config file set up.

**Review by remicousin — APPROVED** [review]

**Comment by drewmresnick** [comment]
So to finish this PR, I will update the file structure so that there is only a subdirectory for resolution if we have something regridded (such as the r0p25 subd) but will move forward assuming that the original resolution will work. 

I am going to edit all of the config files to reflect the changes since so far I only applied it to one config file.

Also @remicousin I know you said we could remove the regridding function but maybe it makes sense to keep it in the file just in case we ever need it? I can just make it so that the conversion to zarr will never attempt to regrid by default. But at least in case someone ever needs it, it could still be there.

**Comment by remicousin** [comment]
yes I think we can probably leave the regridding function invokable by the enactstozarr code.

**Comment by drewmresnick** [comment]
@remicousin I know you already approved this PR but I am wondering if you wanted to look at the last changes I pushed according to the comments that were made, before I merge this?

File: enacts/enactstozarr.py
Comment (remicousin) [inline enacts/enactstozarr.py:39]: @aaron-kaplan should we use raise warnings for this kind of stuff? not sure what that entails.

File: enacts/enactstozarr.py
Comment (remicousin) [inline enacts/enactstozarr.py:58]: So I believe if you literally define it as 
null
in the config file, then it becomes None when you read the config file and assigne to ZARR_RESOLUTION and I assume that's probably better than using 0 (even though 0 is a good choice since it's impossible to have such a resolution but still).

File: enacts/enactstozarr.py
Comment (remicousin) [inline enacts/enactstozarr.py:28]: For the record, I would create an issue about regridding, unless already there but I don't think so, including the TODO comment below and including the desire to generalize to other dimensions.

File: enacts/config-nma.yaml
Comment (remicousin) [inline enacts/config-nma.yaml:28]: Should not the chunks vary with resolution? since we want to strike a balance between number of chunks and their size? and avoid chunks of very different sizes? How to deal with that? Maybe leave a comment about it here so that someone who's going to regrid doesn't forget to think about their chunking?

**Review by remicousin — CHANGES_REQUESTED** [review]
Not sure how to unapprove... so posting an additional empty review requesting changes to see if that would change my approval status.

File: enacts/enactstozarr.py
Comment (drewmresnick) [reply enacts/enactstozarr.py:58]: Ah yes thank you, I was confused as to why defining it as `None` was not working..

File: enacts/enactstozarr.py
Comment (drewmresnick) [reply enacts/enactstozarr.py:28]: an issue for improving the method used for the regridding?

Comment (remicousin) [reply enacts/enactstozarr.py:28]: yes. so that it is on the "issues" radar and so that whoever is going to tackle it is going to be reminded there is this piece of code here.

File: enacts/config-gmet.yaml
Comment (aaron-kaplan) [reply enacts/config-gmet.yaml:29]: This is equivalent to
```
vars:
    precip:
        - Ghana/version3/GHANA_ENACTS/ALL/mrg_precip_daily/
        - Ghana/version3/GHANA_ENACTS/ALL/mrg_precip_daily/
        - precip
    tmax:
        ...
```
Either is acceptable, but in this case I have a slight preference for breaking it up, because the line is rather long and hard to scan this way.

File: enacts/config-nma.yaml
Comment (aaron-kaplan) [reply enacts/config-nma.yaml:28]: Yes

Comment (drewmresnick) [reply enacts/config-nma.yaml:28]: I do think the idea would be to change the chunks at different resolutions - however is it so simple a concept that we are expecting users to know what sizes to be chunking the data? I have not been part of the process for deciding on sizes for chunking so I cannot say for sure, though from what I know so far it seems like something of a case-by-case basis when it comes to deciding on sizes. 

I can at least make a note that says something like: 
"Be aware that changes to resolution may also require different chunking sizes. if regridding, consider updating chunk sizes"

I guess this also brings up the question: should I change the chunk sizes for the Ethiopia maproom, since we are using the native resolution now instead of the regridded data?

File: enacts/enactstozarr.py
Comment (aaron-kaplan) [reply enacts/enactstozarr.py:39]: If we go with the suggestion I made [elsewhere](https://github.com/iridl/python-maprooms/pull/277/files/3251997db01508771fb8000718d78c54db9c284b#r1091296468) , `resolution` will be null unless the config author intended to regrid it. In that case I don't think a warning is necessary.

File: enacts/config-nma.yaml
Comment (aaron-kaplan) [reply enacts/config-nma.yaml:28]: If it's fast enough, then it's good enough for now. Someday we'll do some experimentation to identify good rules for picking a chunking, and hopefully automate it.

Comment (remicousin) [reply enacts/config-nma.yaml:28]: For now the "users" are just us really so I think just a comment to remind the next "user" would do. Plus all that is temporary as well as Aaron mentioned.

**Comment by drewmresnick** [comment]
@remicousin I applied the changes you had suggested and based on comments you and Aaron made so it should be ready for a final review now

File: enacts/config-sng.yaml
Comment (remicousin) [inline enacts/config-sng.yaml:39]: here it's not going to work out: ANACIM/anacim_v3/ALL_block2/ need be either with the paths, or with the vars.

File: enacts/enactstozarr.py
Comment (remicousin) [reply enacts/enactstozarr.py:64]: or generally speaking is there anything that checks that we are not going to overwrite something pre-existing?

**Review by remicousin — CHANGES_REQUESTED** [review]

**Comment by remicousin** [comment]
As we work on this, I am now seeing two other problems that will need to be addressed but maybe not in this PR.

One is that we are ignoring the case that zarrifies dekadal data (used by monthly maproom). I must have run it when at ZMD to generate the zarr files for their monthly version but I don't quite recall what I used and I don't see how set_up_dims could set the time correctly for dekads there.

The 2nd problem is that set_up_dims is pretty optimistic on how the ENACTS data files names are with a specific split made.

File: enacts/enactstozarr.py
Comment (remicousin) [inline enacts/enactstozarr.py:71]: AH! Actually: this is the dekad data case... that's why there were 2 set_up_dims... One came from enactstozarr working on daily data, the other from dekadal...

:(

Sorry Drew, I am afraid you are going to have to reconcile this. Maybe 2 functions: set_up_daily_dims and set_up_dekad_dims and a new argument to convert to tell which one to use?

Then you probably want the config files that use these to be set up like you're setting up the daily data. I think only kmd and zmd have it. And then monthly/maproom.py and layout.py will have to use the right CONFIG info...

Sorry I didn't see that earlier...

File: enacts/enactstozarr.py
Comment (drewmresnick) [reply enacts/enactstozarr.py:64]: It looks to me like I forgot to add the `.zarr` at the end of this line of code. I would think that if I added the file type definition back to this line of code we would not run into an issue here where the data would be overwritten even if the output is to the same directory. But I can also look into how to check if a file already exists under the same name and if so, to throw an error.

File: enacts/enactstozarr.py
Comment (drewmresnick) [reply enacts/enactstozarr.py:71]: Okay, yes in this case I can resolve the other things and then maybe you can take over this part.

File: enacts/config-gmet.yaml
Comment (drewmresnick) [reply enacts/config-gmet.yaml:37]: I can do this yes. I will wait for your response on the other conversation regarding the structure of the config file before I do anything about this part

File: enacts/config-nma.yaml
Comment (remicousin) [reply enacts/config-nma.yaml:52]: I think it makes sense to reproduce the path structure from the nc files to the zarr ones. In that case we need only nc_path and zarr_path to indicate the top part of the path that changes.

**Comment by drewmresnick** [comment]
@remicousin just finished my sanity check, the data looks to be saving correctly. :) I got nervous because something went wrong and one of the zarr files I made was empty when I tried to use it, but now the files I re-made are filled with the data so I think there was something I did on my end when trying to generate data that messed up the conversion.

**Comment by drewmresnick** [comment]
@remicousin can you look at the commit I just pushed to see if this is what you had in mind for the file paths defs?

As per your second comment, are you asking me to update the dictionary as you suggested where the output file path by default is none, meaning that the input and output file dirs are the same?

**Comment by drewmresnick** [comment]
@remicousin I think I am understanding more now what you are asking, are you hoping it would be like this:

nc_path: top level path to where nc files are
zarr_path: top level path to where zarr files will go

vars:
    var_name:
        - common file path 
        - alternate output file path (if not identical to the common file path)
        - var name in data

**Comment by remicousin** [comment]
yes that is what I had in mind. Please apply to all other countries.

**Comment by drewmresnick** [comment]
@remicousin I added the change to the remaining config files

**Review by remicousin — APPROVED** [review]
I approve but we can wait for Aaron's

**Comment by drewmresnick** [comment]
@aaron-kaplan when you have some time could you do a final look through of this PR so we can merge it?

**Review by aaron-kaplan — APPROVED** [review]
Just minor comments.

**Comment by drewmresnick** [comment]
@remicousin this branch is merged now for your use.

File: enacts/enactstozarr.py
Comment (aaron-kaplan) [inline enacts/enactstozarr.py:50]: This doesn't look right. Did you test?

Comment (drewmresnick) [reply enacts/enactstozarr.py:50]: @aaron-kaplan No I did not test because of how long it takes for the code to run, and I thought it was trivial. Rémi is going to open a new PR to continue work on this so should he just correct it in his PR? or should I open a quick PR to fix the issue?

Sorry about that..

Comment (aaron-kaplan) [reply enacts/enactstozarr.py:50]: Not urgent, since this isn't being deployed anywhere.

Comment (drewmresnick) [reply enacts/enactstozarr.py:50]: Okay, then when @remicousin opens his PR to continue working on this, he can make the change - just for clarification, when you suggested this

`input_path, output_path, var_name = CONFIG['vars'][variable]`

Were you implying making three separate variables directly calling the config var such as ex:

`var_name = CONFIG['vars'][variable][2]`

Comment (aaron-kaplan) [reply enacts/enactstozarr.py:50]:
```
input_path, output_path, var_name = CONFIG['vars'][variable]

```
is a more concise way of writing
```
input_path = CONFIG['vars'][variable][0]
output_path = CONFIG['vars'][variable][1]
var_name = CONFIG['vars'][variable][2]
```

Comment (drewmresnick) [reply enacts/enactstozarr.py:50]: Oh I see, so then you do not have to put the [0], [1], [2] at the end of each variable. Thank you for the clarification. :)

**Pull request merged**
